pkg/kubectl/cmd/clusterinfo_dump.go

```diff
@@ -0,0 +1,168 @@
+			cmdutil.CheckErr(dumpClusterInfo(f, cmd, args, cmdOut))
+		},
+	}
+	cmd.Flags().String("output-directory", "", "Where to output the files.  If empty or '-' uses stdout, otherwise creates a directory hierarchy in that directory")
+	cmd.Flags().StringSlice("namespaces", []string{api.NamespaceSystem}, "A comma separated list of namespaces to dump.")
+	cmd.Flags().Bool("all-namespaces", false, "If true, dump all namespaces.  If true, --namespaces is ignored.")
+	return cmd
+}
+
+const (
+	dumpLong = `
+Dumps cluster info out suitable for debugging and diagnosing cluster problems.  By default, dumps everything to
+stdout. You can optionally specify a directory with --output-directory.  If you specify a directory, kubernetes will
+build a set of files in that directory.  By default only dumps things in the 'kube-system' namespace, but you can
+switch to a different namespace with the --namespaces flag, or specify --all-namespaces to dump all namespaces.
+`
+
+	dumpExample = `  # Dump current cluster state to stdout
+  kubectl clusterinfo dump
+  
+  # Dump current cluster state to /tmp
+  kubectl clusterinfo dump --output-directory=/tmp
+  
+  # Dump all namespaces to stdout
+  kubectl clusterinfo dump --all-namespaces
+  
+  # Dump a set of namespaces to /tmp
+  kubectl clusterinfo dump --namespaces default,kube-system --output-directory=/tmp`
+)
+
+func getWriter(cmd *cobra.Command, out io.Writer, filename string) io.Writer {
+	dir := cmdutil.GetFlagString(cmd, "output-directory")
+	if len(dir) == 0 || dir == "-" {
+		return out
+	}
+	fullFile := path.Join(dir, filename)
+	parent := path.Dir(fullFile)
+	cmdutil.CheckErr(os.MkdirAll(parent, 0755))
+
+	file, err := os.Create(path.Join(dir, filename))
+	cmdutil.CheckErr(err)
+	return file
+}
+
+func dumpClusterInfo(f *cmdutil.Factory, cmd *cobra.Command, args []string, out io.Writer) error {
+	var c *unversioned.Client
+	var err error
+	if c, err = f.Client(); err != nil {
+		return err
+	}
+	printer, _, err := kubectl.GetPrinter("json", "")
+	if err != nil {
+		return err
+	}
+
+	nodes, err := c.Nodes().List(api.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	if err := printer.PrintObj(nodes, getWriter(cmd, out, "nodes.json")); err != nil {
+		return err
+	}
+
+	var namespaces []string
+	if cmdutil.GetFlagBool(cmd, "all-namespaces") {
+		namespaceList, err := c.Namespaces().List(api.ListOptions{})
+		if err != nil {
+			return err
+		}
+		for ix := range namespaceList.Items {
+			namespaces = append(namespaces, namespaceList.Items[ix].Name)
+		}
+	} else {
+		namespaces = cmdutil.GetFlagStringSlice(cmd, "namespaces")
+	}
+	for _, namespace := range namespaces {
+		events, err := c.Events(namespace).List(api.ListOptions{})
+		if err != nil {
+			return err
+		}
+		if err := printer.PrintObj(events, getWriter(cmd, out, path.Join(namespace, "events.json"))); err != nil {
+			return err
+		}
+
+		rcs, err := c.ReplicationControllers(namespace).List(api.ListOptions{})
```

Comment:
If we want to support version skewed use cases, we'll need to get the supported resources back:
```go
resourcesList, err := client.Discovery().ServerResources()
```

and then check if each resource is supported before getting it, for example,  

```go
if contains(resourcesList, v1beta1.SchemeGroupVersion.WithResource("deployments")) {
  // Then we can get deployments 
}
```

Ref https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/run.go#L157-L176